components/movementsensor/mpu6050/mpu6050.go

```diff
@@ -0,0 +1,291 @@
+
+	// The scale is +/- 2G's, but our units should be mm/sec/sec.
+	maxAcceleration := 2.0 * 9.81 /* m/sec/sec */ * 1000.0 /* mm/m */
+	return r3.Vector{
+		X: setScale(x, maxAcceleration),
+		Y: setScale(y, maxAcceleration),
+		Z: setScale(z, maxAcceleration),
+	}
+}
+
+func (mpu *mpu6050) AngularVelocity(ctx context.Context, extra map[string]interface{}) (spatialmath.AngularVelocity, error) {
+	mpu.mu.Lock()
+	defer mpu.mu.Unlock()
+
+	// The gyroscope measurements are in registers 67 to 72 (2 bytes for each axis).
+	rawData, err := mpu.readBlock(ctx, 67, 6)
+	if err != nil {
+		return spatialmath.AngularVelocity{}, err
+	}
+	return toAngularVelocity(rawData), nil
+}
+
+func (mpu *mpu6050) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	return r3.Vector{}, movementsensor.ErrMethodUnimplementedLinearVelocity
+}
+
+func (mpu *mpu6050) Orientation(ctx context.Context, extra map[string]interface{}) (spatialmath.Orientation, error) {
+	return nil, movementsensor.ErrMethodUnimplementedOrientation
+}
+
+func (mpu *mpu6050) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	return 0, movementsensor.ErrMethodUnimplementedCompassHeading
+}
+
+func (mpu *mpu6050) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	return geo.NewPoint(0, 0), 0, movementsensor.ErrMethodUnimplementedPosition
+}
+
+func (mpu *mpu6050) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (mpu *mpu6050) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	mpu.mu.Lock()
+	defer mpu.mu.Lock()
+
+	readings := make(map[string]interface{})
+	// Instead of this function calling AngularVelocity, we get an atomic snapshot of all the
+	// values together, rather than the angular velocity at one time and everything else at a
+	// slightly different time.
+
+	// We get 2 bytes each for 3 axes of acceleration, 1 temperature, and 3 axes of rotation.
+	rawData, err := mpu.readBlock(ctx, 59, 14)
+	if err != nil {
+		return readings, err
+	}
+
+	readings["acceleration"] = toLinearAcceleration(rawData[0:6])
+
+	// Taken straight from the MPU6050 register map. Yes, these are weird constants.
+	temp := toSignedValue(rawData[6:8])
+	readings["temperature_celsius"] = float64(temp)/340 + 36.53
+
+	readings["angular_velocity"] = toAngularVelocity(rawData[8:14])
+
+	return readings, nil
+}
+
+func (mpu *mpu6050) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
+	return &movementsensor.Properties{
+		AngularVelocitySupported: true,
+	}, nil
+}
+
+func (mpu *mpu6050) Close(ctx context.Context) {
+	mpu.mu.Lock()
+	defer mpu.mu.Lock()
+
+	// Set the Sleep bit (bit 6) in the power control register (register 107).
+	err := mpu.writeByte(ctx, 107, 1<<6)
```

Comment:
Do you want to close the i2c handle here? Or does writing a sleep to the chip do that on the firmware of the chip itself?

Reply:
The implementation of `writeByte` opens the handle, writes the byte, and closes it again. That way, the I2C bus is totally free for others to use almost all of the time, and we don't need to do anything special here.